### ISSUE TITLE
FP: `needless_collect` for `HashMap` and `BTreeMap`

### ISSUE THREAD

**Issue by mgacek8** [rust-lang/rust-clippy#7164]
Lint name: `needless_collect`

I tried this code:

```rust
use std::collections::{HashMap, BTreeMap};

fn main() {
    let sample = [1; 5];
    let a = sample.iter().map(|x| (x, x)).count();
    let b = sample.iter().map(|x| (x, x)).collect::<HashMap<_, _>>().len();
    let c = sample.iter().map(|x| (x, x)).collect::<BTreeMap<_, _>>().len();
    println!("{}, {}, {}", a, b, c);
}
```

I expected to see this happen: No lint triggered.
Instead, this happened: 
Warning from clippy:
```

warning: avoid using `collect()` when not needed
 --> src/main.rs:6:43
  |
6 |     let b = sample.iter().map(|x| (x, x)).collect::<HashMap<_, _>>().len();
  |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `count()`
  |
  = note: `#[warn(clippy::needless_collect)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect

warning: avoid using `collect()` when not needed
 --> src/main.rs:7:43
  |
7 |     let c = sample.iter().map(|x| (x, x)).collect::<BTreeMap<_, _>>().len();
  |                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: replace with: `count()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_collect
  ```
After applying suggestion from clippy, the results are different. 
Before applying the suggestion: 1, after: 5.

[playgroud link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6201852b51608ca89cd645f671b5ea82)

### Meta

- `clippy`: 0.1.53 (2021-05-03 716394d)
- `rustc`: 1.54.0-nightly (2021-05-03 716394d6581b60c75cfd)
